ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,282 @@
+use std::collections::BTreeMap;
+
+use futures::{stream, StreamExt, TryStreamExt};
+
+use crate::{
+    error::Error,
+    ff::{boolean::Boolean, boolean_array::BooleanArray, ArrayAccess},
+    helpers::TotalRecords,
+    protocol::{
+        boolean::step::EightBitStep,
+        context::{
+            dzkp_validator::{validated_seq_join, DZKPValidator, TARGET_PROOF_SIZE},
+            Context, DZKPUpgraded, MaliciousProtocolSteps, ShardedContext, UpgradableContext,
+        },
+        hybrid::step::{AggregateReportsStep, HybridStep},
+        ipa_prf::boolean_ops::addition_sequential::integer_add,
+        BooleanProtocols,
+    },
+    report::hybrid::{AggregateableHybridReport, PrfHybridReport},
+    secret_sharing::replicated::semi_honest::AdditiveShare as Replicated,
+};
+
+enum MatchEntry<BK, V>
+where
+    BK: BooleanArray,
+    V: BooleanArray,
+{
+    Empty,
+    Single(AggregateableHybridReport<BK, V>),
+    Pair(
+        AggregateableHybridReport<BK, V>,
+        AggregateableHybridReport<BK, V>,
+    ),
+    MoreThanTwo,
+}
+
+impl<BK, V> MatchEntry<BK, V>
+where
+    BK: BooleanArray,
+    V: BooleanArray,
+{
+    pub fn add_report(&mut self, new_report: AggregateableHybridReport<BK, V>) {
+        match self {
+            Self::Empty => *self = Self::Single(new_report),
+            Self::Single(old_report) => {
+                *self = Self::Pair(old_report.clone(), new_report);
+            }
+            Self::Pair { .. } | Self::MoreThanTwo => *self = Self::MoreThanTwo,
+        }
+    }
+}
+
+/// This function takes in a vector of `PrfHybridReports`, groups them by the oprf of the `match_key`,
+/// and collects all pairs of reports with the same `match_key` into a vector of paris (as an array.)
+/// Note that any `match_key` which appears once or more than twice is removed.
+/// An honest report collector will only provide a single impression report per `match_key` and
+/// an honest client will only provide a single conversion report per `match_key`.
+fn group_report_pairs<BK, V>(
+    reports: Vec<PrfHybridReport<BK, V>>,
+) -> Vec<[AggregateableHybridReport<BK, V>; 2]>
+where
+    BK: BooleanArray,
+    V: BooleanArray,
+{
+    let mut reports_by_matchkey: BTreeMap<u64, MatchEntry<BK, V>> = BTreeMap::new();
+
+    for report in reports {
+        let match_entry = reports_by_matchkey
+            .entry(report.match_key)
+            .or_insert(MatchEntry::Empty);
+        match_entry.add_report(report.into());
+    }
+
+    // we only keep the reports from match_keys that provided exactly 2 reports
+    reports_by_matchkey
+        .into_values()
+        .filter_map(|match_entry| match match_entry {
+            MatchEntry::Pair(r1, r2) => Some([r1, r2]),
+            _ => None,
+        })
```

Comment:
if you create a method `MatchEntry.to_pair(self) -> Option<[...; 2]>`, this can be made a one-liner: `filter_map(MatchEntry::to_pair)`